### ISSUE TITLE
feature request: use opam's extra-source and output a shell script to layout the duniverse folder

### ISSUE THREAD

**Issue by hannesm** [ocamllabs/opam-monorepo#242]
At the moment, on `opam-monorepo lock` an opam file is generated which includes:
- all dependencies with strict version bounds
- a set of pin-depends pointing to tarball or git repository
- a set of x-opam-monorepo-duniverse-dirs

The second and third item are mostly the same (are they different? are they used by tooling?).

Instead o a custom opam field, why isn't [`extra-source`](https://opam.ocaml.org/doc/Manual.html#opamsection-extra-sources) used? This would help a lot for a (binary, bit-by-bit) reproducible setup (where the rebuilder does not depend on the opam-monorepo codebase, and all dependencies are tightly tracked). Since extra-source only supports files (i.e. no version control urls), some rewriting would be needed - but all repository hosting platforms support a tar or zip download of a specific git commit.

If we would have extra-source in the generated opam file, an `opam install --download-only` could download the required files for a build, and a shell script could be generated (at `opam-monorepo lock` time) that unpacks these archives into a `duniverse/` layout.

Another minor item is how to lock a version-control remote (see `opam switch export --freeze`), and why only some packages have checksums recorded (maybe an `opam-monorepo lock --freeze --checksums` that retrieves the sources and records checksums would be a good fit).

TL;DR: would you be interested in such a feature? Should this be developed as a new (experimental) subcommand? Are there users of the generated `pin-depens`?

**Comment by hannesm** [comment]
Sorry, I first wanted to thank you so much for this incredible tool. Thanks for developing it, maintaining it and making it freely available. This is a game changer for large (distributed over hundreds of git repositories) codebases. :) 👍🏾

**Comment by hannesm** [comment]
Since there seems to be no activity / interest in this, and I managed to cope with whatever opam-monorepo 0.3.2 does, I'm closing this.
